Github GraphQL Support For On-Prem Github

midnightconman opened the pull request:
This change adds a `github-graphql-endpoint` flag for passing to github graphql client. This mostly fixes 
 github enterprise for tide.

Enhancement on https://github.com/kubernetes/test-infra/pull/11075

components: hook, tide, label_sync, tackle, and prstatus

I am not sure whom to include for approval on this, please make testing and change suggestions.

stevekuznetsov commented:
The robots will assign someone useful, that's their job :)

stevekuznetsov commented:
/ok-to-test

stevekuznetsov commented on prow/prstatus/prstatus.go:
```diff
@@ -37,11 +37,12 @@ import (
-	loginKey       = "login"
+	loginSession          = "github_login"
+	githubEndpoint        = "https://api.github.com"
+	githubGraphQLEndpoint = "https://api.github.com/graphql"
```
Can we import these from somewhere?

stevekuznetsov commented on prow/github/client.go:
```diff
@@ -226,11 +229,14 @@ func NewClient(getToken func() []byte, bases ...string) *Client {
 //   this client to bypass the cache if it is temporarily unavailable.
-func NewDryRunClient(getToken func() []byte, bases ...string) *Client {
+// 'graphqlEndpoint' is the Github GraphQL API (v4) endpoint to use.
+//   The only difference between githubql.NewEnterpriseClient and
```
Not sure we need to keep this comment here

midnightconman replied:
I'll remove them, no problem.

midnightconman commented on prow/flagutil/github.go:
```diff
@@ -49,7 +51,11 @@ func (o *GitHubOptions) AddFlagsWithoutDefaultGithubTokenPath(fs *flag.FlagSet)
 
 func (o *GitHubOptions) addFlags(wantDefaultGithubTokenPath bool, fs *flag.FlagSet) {
 	o.endpoint = NewStrings("https://api.github.com")
+	o.gitEndpoint = "https://github.com"
```
I just merged from parent branch (https://github.com/kubernetes/test-infra/pull/11075), the changes should be merged now.

stevekuznetsov commented:
@midnightconman FYI usually `git rebase` will get you where you want faster than `git merge`

midnightconman commented:
Going to work on these tests... setting DNM until they are working.

/hold

midnightconman commented:
I am waiting on https://github.com/kubernetes/test-infra/pull/11214 to keep working on this.

midnightconman commented:
Hey Guys, starting at the bottom again. This is needed for GitHub Enterprise support. 

As the shurcool library assumes `github.com` if you use `NewClient`, I am using `NewEnterpriseClient`. Please take a look and let me know what you think. It seems the only two apps left not using the centralized `flagutil/github.go` are `label_sync` and `robots/commenter`, so I updated the flags for those.

Thank you again for your time and reviews.

fejta commented on prow/github/client.go:
```diff
@@ -231,14 +231,16 @@ func (c *Client) Throttle(hourlyTokens, burst int) {
-			Timeout:   maxRequestTime,
-			Transport: &oauth2.Transport{Source: newReloadingTokenSource(getToken)},
-		}),
+		gqlc: githubql.NewEnterpriseClient(
```
What are the side effects of doing this for non-enterprise customers? Should pick which to use more carefully?

midnightconman replied:
No side effect, this function wraps `NewClient()` and adds an endpoint url. Here is a link to this function: https://github.com/shurcooL/githubv4/blob/master/githubv4.go#L26

midnightconman replied:
I think the function is poorly named, but it is nice they didn't break signature.

midnightconman commented:
/retest

midnightconman commented:
/retest

midnightconman commented:
/test pull-test-infra-verify-govet

stevekuznetsov commented on prow/prstatus/prstatus.go:
```diff
@@ -35,11 +35,12 @@ import (
-	tokenKey       = "access-token"
-	loginKey       = "login"
+	loginSession    = "github_login"
+	githubEndpoint  = "https://api.github.com"
```
Can we expose these constants in the GitHub package and use them where appropriate? Would hate to typo `graphql` in a flag default...

midnightconman replied:
I'm fine with it, hopefully there won't be any cyclical dependency issues. :crossed_fingers:

midnightconman replied:
This also kinda touches on the shared flagset conversation we had a week or so ago. Do we want to throw these defaults (constants) in `prow/github` and inherit them in `prow/flagutils/github.go` and in these desperate applications? Or do we want to leave these separate?

stevekuznetsov commented:
@fejta do we want to get some traction on #12137 first?

midnightconman commented:
/test pull-test-infra-verify-github-spelling

midnightconman commented:
@stevekuznetsov @fejta - PTAL :smile: 

I added centralized defaults for github api endpoint and github graphql endpoint to `prow/github/types.go`... any application using `prow/github` I changed to use these defaults, anything using `vendor/go-github` is still using static strings.

fejta commented:
/uncc
/cc @cjwagner

stevekuznetsov commented on prow/flagutil/github.go:
```diff
@@ -66,6 +68,10 @@ func (o *GitHubOptions) Validate(dryRun bool) error {
 		}
 	}
 
+	if _, err := url.Parse(o.graphqlEndpoint); err != nil {
```
This one doesn't validate that something _is_ passed

midnightconman replied:
Empty would only happen if someone explicitly specified it, the default is covered by the stringvar flag default.

I also think this parse would fail on empty.

stevekuznetsov replied:
That's the same as in the other place you have written the check, so let's make them identical code.

midnightconman replied:
Ohh, I think I know what you are talking about. I basically copied the existing github endpoint check for each file, so the check logic would be the same. I can update the check logic for github endpoint and graphql endpoint in `flagutil/github.go`, but I would prefer to not update any other files.

This goes back to the flagutil consolidation effort we spoke about a week or two ago. 

Sound good?

stevekuznetsov replied:
I would just like to see the content of `Validate()` be the same for every place we're asking for this as a flag

Noahhoetger2001 replied:
have you tried rerouting it to .shell maybe it might work

Noahhoetger2001 replied:
another thing. what about the parse code being implemented into something else so it can be passed

Noahhoetger2001 replied:
/uncc

Noahhoetger2001 replied:
/uncc
/cc @Noahhoetger2001

midnightconman commented:
@cjwagner @stevekuznetsov - PTAL

I updated flagutil to validate we can't have an empty endpoint. Please let me know if you guys need any other changes or have other suggestions.

Noahhoetger2001 commented:
> Going to work on these tests... setting DNM until they are working.
> 
> /hold
> /echo

midnightconman commented:
Hi Noah, welcome. Let's let Steve and Cole review this code and we can go from there.

/hold cancel

noahwolf2001 commented:
/test

cjwagner commented on prow/flagutil/github.go:
```diff
@@ -61,11 +63,19 @@ func (o *GitHubOptions) addFlags(wantDefaultGitHubTokenPath bool, fs *flag.FlagS
 	for _, uri := range o.endpoint.Strings() {
-		if _, err := url.ParseRequestURI(uri); err != nil {
+		if uri == "" {
+			uri = github.DefaultAPIEndpoint
```
The default is already specified above. If `uri == ""` here then `--github-endpoint` was explicitly set to the empty value which should be an error.

midnightconman replied:
I am using the default in validate also, as some testing skips addFlags step, so this value would be empty. Is it possible to have a follow-up PR for this, I have had a bit of back and forth on this requirement from Steve and Fejta. 

The general consensus was to not couple validation logic across individual components.

cjwagner commented on prow/flagutil/github.go:
```diff
@@ -61,11 +63,19 @@ func (o *GitHubOptions) addFlags(wantDefaultGitHubTokenPath bool, fs *flag.FlagS
 	}
 
+	if o.graphqlEndpoint == "" {
+		o.graphqlEndpoint = github.DefaultGraphQLEndpoint
```
Defaulting is already handled as above.

midnightconman replied:
Same here, I am explicitly defaulting here, as some individual testing implementations don't run `addFlags`, so this value is empty for those tests.

cjwagner commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -55,6 +56,7 @@ func flagOptions() options {
 	flag.StringVar(&o.jenkinsUserName, "jenkins-user-name", "", "Jenkins username.")
 
 	flag.StringVar(&o.githubEndpoint, "github-endpoint", "https://api.github.com", "GitHub's API endpoint.")
+	flag.StringVar(&o.graphqlEndpoint, "graphql-endpoint", "https://api.github.com/graphql", "GitHub's GraphQL API endpoint.")
```
Pleas use the default value constants.

midnightconman replied:
Good catch, switched to use defaults.

midnightconman commented:
@cjwagner - PTAL

Thank you for your reviews :smile:

midnightconman commented:
@stevekuznetsov - Cole approved, can I remove hold, or can you, please? :smile:

cjwagner commented:
/hold cancel

noahwolf2001 commented:
/hold cancel

noahwolf2001 commented:
how long do you think the infra test will be complete
